test: randomized failure injection for Raft-based topology

enaydanov opened the pull request:
The idea of the test is to have a cluster where one node is stressed with injections and failures and the rest of the cluster is used to make progress of the raft state machine.

To achieve this following two lists introduced in the PR:

  - ERROR_INJECTIONS in error_injections.py
  - CLUSTER_EVENTS in cluster_events.py

Each cluster event is an async generator which has 2 yields and should be used in the following way:

   0. Start the generator:
       ```python
       >>> cluster_event_steps = cluster_event(manager, random_tables, error_injection)
       ```

   1. Run the prepare part (before the first yield)
       ```python
       >>> await anext(cluster_event_steps)
       ```

   2.  Run the cluster event itself (between the yields)
       ```python
       >>> await anext(cluster_event_steps)
       ```

   3. Run the check part (after the second yield)
       ```python
       >>> await anext(cluster_event, None)
       ```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8283/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8283/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_random_failures
#### Failed Tests (141/900):
- [test_random_failures[stop_before_streaming-add_new_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_before_streaming_add_new_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_streaming-add_new_node])
- [test_random_failures[stop_after_streaming-stop_non_coordinator_node_gracefully]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_streaming_stop_non_coordinator_node_gracefully_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-stop_non_coordinator_node_gracefully])
- [test_random_failures[stop_after_streaming-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_streaming_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-remove_data_dir_of_dead_node])
- [test_random_failures[stop_after_starting_auth_service-decommission_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_auth_service_decommission_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_auth_service-decommission_node])
- [test_random_failures[stop_after_streaming-restart_non_coordinator_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_streaming_restart_non_coordinator_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-restart_non_coordinator_node])
- [test_random_failures[stop_after_sending_join_node_request-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_sending_join_node_request_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_sending_join_node_request-remove_data_dir_of_dead_node])
- [test_random_failures[stop_after_starting_group0_service-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_group0_service_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_group0_service-remove_data_dir_of_dead_node])
- [test_random_failures[stop_after_bootstrapping_initial_raft_configuration-decommission_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_bootstrapping_initial_raft_configuration_decommission_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_bootstrapping_initial_raft_configuration-decommission_node])
- [test_random_failures[stop_after_sending_join_node_request-remove_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_sending_join_node_request_remove_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_sending_join_node_request-remove_node])
- [test_random_failures[stop_after_bootstrapping_initial_raft_configuration-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8283/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_bootstrapping_initial_raft_configuration_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_bootstrapping_initial_raft_configuration-remove_data_dir_of_dead_node])
#### Build Details:
- Duration: 8 hr 41 min
- Builder: i-0ebe813255e86b1cb (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9517/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9517/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_random_tables
:small_blue_diamond: topology_random_failures/test_random_failures
#### Failed Tests (191/260):
- [test_random_failures[stop_after_bootstrapping_initial_raft_configuration-init_tablet_transfer]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_bootstrapping_initial_raft_configuration_init_tablet_transfer_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_bootstrapping_initial_raft_configuration-init_tablet_transfer])
- [test_random_failures[stop_after_starting_migration_manager-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_migration_manager_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_migration_manager-remove_data_dir_of_dead_node])
- [test_random_failures[stop_after_updating_cdc_generation-insert_records]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_updating_cdc_generation_insert_records_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_updating_cdc_generation-insert_records])
- [test_random_failures[stop_after_sending_join_node_request-restart_coordinator_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_sending_join_node_request_restart_coordinator_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_sending_join_node_request-restart_coordinator_node])
- [test_random_failures[stop_after_setting_mode_to_normal-execute_lwt_transaction]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_setting_mode_to_normal_execute_lwt_transaction_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_setting_mode_to_normal-execute_lwt_transaction])
- [test_random_failures[stop_after_init_of_schema_commitlog-restart_non_coordinator_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_schema_commitlog_restart_non_coordinator_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_schema_commitlog-restart_non_coordinator_node])
- [test_random_failures[stop_after_sending_join_node_request-init_tablet_transfer]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_sending_join_node_request_init_tablet_transfer_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_sending_join_node_request-init_tablet_transfer])
- [test_random_failures[stop_during_gossip_shadow_round-execute_lwt_transaction]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_during_gossip_shadow_round_execute_lwt_transaction_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_during_gossip_shadow_round-execute_lwt_transaction])
- [test_random_failures[stop_after_init_of_system_ks-remove_data_dir_of_dead_node]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_system_ks_remove_data_dir_of_dead_node_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_system_ks-remove_data_dir_of_dead_node])
- [test_random_failures[stop_after_setting_mode_to_normal-execute_lwt_transaction]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9517/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_setting_mode_to_normal_execute_lwt_transaction__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_setting_mode_to_normal-execute_lwt_transaction])
#### Build Details:
- Duration: 1 hr 8 min
- Builder: spider3.cloudius-systems.com

temichus commented on test/topology_random_failures/suite.yaml:
```diff
@@ -0,0 +1,11 @@
+type: Topology
+pool_size: 4
```
because now test.py parallel tests run per file. pool size can be 1 now

enaydanov replied:
done

temichus commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,167 @@
+
+@pytest.mark.usefixtures("four_nodes_cluster")
+@pytest.mark.asyncio
+@pytest.mark.slow
```
is plugin https://pypi.org/project/pytest-skip-slow/ is installed for test.py to skip theses tests in CI? probably not

enaydanov replied:
removed the marker

temichus commented:
code looks good and structured.
questions
1) why some of the tests failed, is these failures expected? (if yes should we disable this test for now?)
2) how much CI run for ScyllaDB PRs will take after we merge this?

temichus commented on test/topology_experimental_raft/test_random_tables.py:
how is this file renaming connected with this PR?

enaydanov replied:
Oups. No connection, moved it by accident. 

Reverted.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9723/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9723/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
#### Failed Tests (9/80):
- [test_random_failures[stop_before_becoming_raft_voter-add_new_node].debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_before_becoming_raft_voter_add_new_node__debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_becoming_raft_voter-add_new_node].debug.10)
- [test_random_failures[stop_after_updating_cdc_generation-kill_non_coordinator_node].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_updating_cdc_generation_kill_non_coordinator_node__debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_updating_cdc_generation-kill_non_coordinator_node].debug.4)
- [test_random_failures[stop_after_starting_repair-restart_non_coordinator_node].debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_repair_restart_non_coordinator_node__debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_repair-restart_non_coordinator_node].debug.6)
- [test_random_failures[stop_after_init_of_schema_commitlog-remove_data_dir_of_dead_node].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_schema_commitlog_remove_data_dir_of_dead_node__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_schema_commitlog-remove_data_dir_of_dead_node].debug.7)
- [test_random_failures[stop_after_streaming-restart_non_coordinator_node].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_streaming_restart_non_coordinator_node__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-restart_non_coordinator_node].debug.7)
- [test_random_failures[stop_after_starting_auth_service-add_new_node].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_auth_service_add_new_node__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_auth_service-add_new_node].debug.7)
- [test_random_failures[stop_before_streaming-restart_coordinator_node].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_before_streaming_restart_coordinator_node__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_streaming-restart_coordinator_node].debug.7)
- [test_random_failures[stop_before_streaming-init_tablet_transfer].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_before_streaming_init_tablet_transfer__debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_streaming-init_tablet_transfer].debug.8)
- [test_random_failures[stop_after_saving_tokens-remove_data_dir_of_dead_node].debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9723/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_saving_tokens_remove_data_dir_of_dead_node__debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_saving_tokens-remove_data_dir_of_dead_node].debug.9)
#### Build Details:
- Duration: 1 hr 45 min
- Builder: spider7.cloudius-systems.com

enaydanov commented:
> code looks good and structured. questions:
> 
>     1. why some of the tests failed, is these failures expected? (if yes should we disable this test for now?)
> 
>     2. how much CI run for ScyllaDB code change PRs will take after we merge this?
> 
>     3. could you please provide the result of the full test run(all combinations)?

1. Some of failures are known issues, added a commit to the PR with skipping of these tests. Other required some investigations (in progress.)
2. This is not measured yet.
3. Added to the PR description.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9850/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9850/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
#### Failed Tests (9/160):
- [test_random_failures[stop_after_streaming-insert_records].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_streaming_insert_records__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-insert_records].debug.1)
- [test_random_failures[stop_after_init_of_system_ks-init_tablet_transfer].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_system_ks_init_tablet_transfer__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_system_ks-init_tablet_transfer].debug.1)
- [test_random_failures[stop_after_sending_join_node_request-drop_table].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_sending_join_node_request_drop_table__debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_sending_join_node_request-drop_table].debug.4)
- [test_random_failures[stop_after_starting_auth_service-add_new_keyspace].debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_auth_service_add_new_keyspace__debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_auth_service-add_new_keyspace].debug.5)
- [test_random_failures[stop_after_starting_gossiping-drop_udt].debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_gossiping_drop_udt__debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_gossiping-drop_udt].debug.6)
- [test_random_failures[stop_after_starting_commitlog-update_record].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_starting_commitlog_update_record__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_commitlog-update_record].debug.7)
- [test_random_failures[stop_after_init_of_system_ks-decommission_node].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_system_ks_decommission_node__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_system_ks-decommission_node].debug.7)
- [test_random_failures[stop_after_init_of_system_ks-update_record].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_system_ks_update_record__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_system_ks-update_record].debug.7)
- [test_random_failures[stop_after_init_of_schema_commitlog-restart_coordinator_node].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9850/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_init_of_schema_commitlog_restart_coordinator_node__debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_init_of_schema_commitlog-restart_coordinator_node].debug.8)
#### Build Details:
- Duration: 2 hr 48 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9903/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9903/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
#### Failed Tests (2/400):
- [test_random_failures[stop_after_updating_cdc_generation-kill_coordinator_node].debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9903/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_updating_cdc_generation_kill_coordinator_node__debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_updating_cdc_generation-kill_coordinator_node].debug.6)
- [test_random_failures[stop_after_updating_cdc_generation-kill_coordinator_node].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9903/testReport/junit/%28root%29/test_random_failures/test_random_failures_stop_after_updating_cdc_generation_kill_coordinator_node__debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_updating_cdc_generation-kill_coordinator_node].debug.8)
#### Build Details:
- Duration: 4 hr 46 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9922/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9922/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9922/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10273/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2252/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9922/)
#### Failed Tests (1/31820):
- [test_repair_kill_1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9922/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_1)
#### Build Details:
- Duration: 9 hr 21 min
- Builder: spider3.cloudius-systems.com

enaydanov commented:
Looks like we got https://github.com/scylladb/scylladb/issues/19223 here.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9981/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9981/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9981/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10309/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2292/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9981/)
#### Build Details:
- Duration: 9 hr 31 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10004/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10004/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10004/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10327/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10004/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2309/)
#### Build Details:
- Duration: 9 hr 27 min
- Builder: spider3.cloudius-systems.com

temichus reviewed (approved):
so, it is green. And last commit shows mechanism how to disabling tests in case of flakiness/issues
LGTM

enaydanov commented:
Full matrix run passed:

https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2316/testReport/(root)/test_random_failures/

```
00:09:00.320  + bash -c 'set -o pipefail; ./tools/toolchain/dbuild -v `pwd`:/scylladir -v /jenkins/coredumps:/jenkins/coredumps -- ./test.py  --tmpdir=/scylladir/testlog/x86_64 --artifacts_dir_url https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2316/artifact --verbose --repeat 1 topology_random_failures --mode=debug 2>&1 | tee tests_console_output_x86_64.txt'
00:09:00.321  Found 1 tests.
00:09:00.927  ================================================================================
00:09:00.927  [N/TOTAL]   SUITE    MODE   RESULT   TEST
00:09:00.927  ------------------------------------------------------------------------------
04:11:21.060  [1/1]      topology_random_failures  debug  [ PASS ] topology_random_failures.test_random_failures.1 14509.29s
04:11:21.060  ------------------------------------------------------------------------------
04:11:21.060  CPU utilization: 3.8%
```

kbr-scylla commented:
I think we should resolve the known issues so that we can merge the test without needing to skip injections.

I commented on each issue you listed in the cover letter.

BTW. when opening such issues, please assign someone, or add triage label -- otherwise nobody will look at it (and I see these issues hang since May) (actually I'm not sure if adding triage is enough; best to assign someone)

kbr-scylla commented:
@temichus my problem is that these issues will not be fixed because they describe expected behavior of the system.
I can close them as "not planned". Then what?

temichus commented:
> @temichus my problem is that these issues will not be fixed because they describe expected behavior of the system. I can close them as "not planned". Then what?

so, if "not planned" we will check for a solution to remove these tests from run, instead of "Skip" them

kostja commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,631 @@
+                                   error_injection: str) -> AsyncIterator[None]:
+    yield
+
+    LOGGER.info("Restart the coordinator node")
```
Can we have a more reliable way to detect the raft leader here?
cc @kbr-scylla

enaydanov replied:
I think we can use this function:

https://github.com/scylladb/scylladb/blob/c11f2c9bcdffe00622a5fea0e2171d53fe361b68/test/topology/util.py#L363-L371

enaydanov replied:
Done

kostja replied:
@kbr-scylla ^^

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10389/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10389/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10389/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2592/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10611/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10389/)
#### Build Details:
- Duration: 3 hr 17 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10401/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10401/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10401/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2604/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10623/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10401/)
#### Build Details:
- Duration: 3 hr 31 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10824/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10824/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10824/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10982/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2960/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10824/)
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10852/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10852/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10852/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11013/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2990/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10852/)
#### Build Details:
- Duration: 3 hr 12 min
- Builder: spider1.cloudius-systems.com

enaydanov commented:
Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2442/testReport/(root)/test_random_failures/

@kbr-scylla can you review/approve it?

cc @kostja @temichus

kbr-scylla commented:
> Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2442/testReport/(root)/test_random_failures/

Does the PR description and linked sheet have to be updated?

In fact the sheet should probably not be included in the PR description, or the known issues -- all of that is going to be made permanent part of Scylla's `git log`.

kbr-scylla commented on test/pylib/scylla_cluster.py:
```diff
@@ -1682,6 +1699,13 @@ async def _server_get_sstables_disk_usage(self, request: aiohttp.web.Request) ->
         return self.cluster.get_sstables_disk_usage(server_id, data["keyspace"], data["table"])
 
+    async def _server_get_process_status(self, request: aiohttp.web.Request) -> Optional[str]:
+        """Stop a server. No-op if already stopped."""
```
docstring wrong

kbr-scylla commented on test/topology/util.py:
```diff
@@ -366,11 +367,18 @@ async def get_coordinator_host(manager: ManagerClient) -> ServerInfo:
+        with suppress(Exception):
+            if await manager.get_host_id(s_info.server_id) == coordinator_host_id:
+                return s_info
+    raise AssertionError(f"There is no host with id={coordinator_host_id} in the cluster")
```
nit: the previous error provided more potentially useful details -- the list of host IDs present in the cluster.

enaydanov replied:
No, it's for suppressing Exception from get_host_id() call:

```python
    async def get_host_id(self, server_id: ServerNum) -> HostID:
        """Get local host id of a server"""
        try:
            host_id = await self.client.get_json(f"/cluster/host-id/{server_id}")
        except Exception as exc:
            raise Exception(f"Failed to get local host id address for server {server_id}") from exc
        return HostID(host_id)
```

kbr-scylla replied:
I don't understand your answer. My complaint is that the message you removed:
```
        f"Node with host id {coordinator_host_id} was not found in cluster host ids {list(server_id_maps.keys())}"
```

was more rich in details compared to the new message
```
f"There is no host with id={coordinator_host_id} in the cluster"
```

enaydanov replied:
OK, got it.

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -180,15 +180,18 @@ async def replicas(self) -> int:
 
-    async def running_servers(self) -> list[ServerInfo]:
-        """Get List of server info (id and IP address) of running servers"""
+    async def running_servers(self, *server_ids: int) -> list[ServerInfo]:
```
The caller of this function can intersect the returned list with their chosen filter. No need to complicate the function itself.

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2129,6 +2130,10 @@ future<> gossiper::do_shadow_round(std::unordered_set<gms::inet_address> nodes,
         auto start_time = clk::now();
         std::list<gms::gossip_get_endpoint_states_response> responses;
+
+        utils::get_local_injector().inject("stop_during_gossip_shadow_round",
```
note that here the shadow round hasn't even really started yet, we didn't send any messages or update any state. So this is basically "stop **before** shadow round".

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2024,6 +2034,9 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
 
     set_mode(mode::NORMAL);
 
+    utils::get_local_injector().inject("stop_after_setting_mode_to_normal",
```
Two injections with the same name?

enaydanov replied:
One for Raft and another one for gossip.

enaydanov replied:
No, there is how it works:

https://github.com/scylladb/scylladb/blob/15f8046fcb771e496d7e0db36ff2051c100d2593/utils/error_injection.hh#L450-L474

In case of this error injection it's not possible to reach both points in one run anyway.

kostja replied:
I'd still disambiguate this @enaydanov . Why do you need two separate injections, imagine the maintenance of this test, one would have to wonder which injection worked this time when debugging this.

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -1611,6 +1611,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                             "bootstrap: insert tokens and CDC generation data (UUID: {})", gen_uuid);
                         co_await update_topology_state(std::move(guard), {std::move(mutation), builder.build()}, reason);
+
+                        co_await utils::get_local_injector().inject("stop_after_updating_cdc_generation",
```
"stop" -> "pause"

enaydanov replied:
I see "topology_coordinator_pause_before_processing_backlog" injection in the same file, so changed it to "topology_coordinator_pause_after_updating_cdc_generation". But I'll keep "stop_after_updating_cdc_generation" as a name for error injection in the matrix.

kbr-scylla commented on test/topology_random_failures/conftest.py:
```diff
@@ -0,0 +1,28 @@
+
+def pytest_addoption(parser: pytest.Parser) -> None:
+    # Limit for number of tests in topology_experimental_raft/test_random_failures.  Can be set in pytest.ini
+    parser.addoption('--raft-failure-injections-tests-count', action='store', default=None, type=int,
```
How is this different from `--repeat`?

enaydanov replied:
It's a number of random tests from the full matrix to run.

kbr-scylla replied:
If I understand correctly, when we run `pytest ... test_random_failures.py`, it will randomly select an `error_injection` and a `cluster_event` and run the `test_random_failures` function with those selected values.

If I increase this parameter to 10 in `pytest.ini`, it will select 10 random injections and 10 random events, and run `test_random_failures` for each pair of selection (?) so 10 times in total.

I will achieve the exact same effect by using `--repeat 10`, no?

So I don't understand the difference.

cc @kostja

kbr-scylla replied:
In general I'm afraid that you're adding a bunch of new parameters that nobody will know they exist (maybe I will know for a week and then forget), so nobody will use them.

We should not add parameters that are not strictly needed.

enaydanov replied:
These are not framework-wide parameters. They are just for this suite.

If `--raft-failure-injections-tests-count=10` then we will run 10 random pairs injections and events without repeats. `--repeat=10` will not run just unique pairs in general.

kbr-scylla replied:
My question still applies. Honestly tell me, who will use these parameters once the code is merged into `master`? If you expect core devs to use it -- they won't.

Maybe you are planning to create a new Jenkins job (parallel to our regular CI runs) that would run this suite with custom parameters in regular fashion? Maybe that would give reason for these parameters to exist.

I understand need for `seed` -- you need it to reproduce a failure.

But what about `index`? What do we need it for?

kostja replied:
@enaydanov I think this can be options in the beginning of the test file or test arguments with default values. I don't think anyone who understand what this test does will have a desire to configure it through pytest.ini - it can be well "configured" by patching the test itself.

kbr-scylla commented on test/topology_random_failures/conftest.py:
```diff
@@ -0,0 +1,28 @@
+    parser.addoption('--raft-failure-injections-seed', action='store', default=None, type=int,
+                     help='seed for RAFT randomized failure injections')
+    parser.addoption('--raft-errors-last-index', action='store', default=None, type=int,
+                     help='last index for RAFT failure injection types')
```
Needs to define what "index" is.

kbr-scylla commented on test/topology_random_failures/conftest.py:
```diff
@@ -0,0 +1,28 @@
+    parser.addoption('--raft-errors-last-index', action='store', default=None, type=int,
+                     help='last index for RAFT failure injection types')
+    parser.addoption('--raft-events-last-index', action='store', default=None, type=int,
+                     help='last index for event types during RAFT failure injections')
```
Needs to define what "index" is.

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+# Copyright (C) 2023-present ScyllaDB
```
2024

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+from test.topology.util import get_coordinator_host, get_non_coordinator_host
+from test.topology_random_failures.error_injections import ERROR_INJECTIONS
+
+if TYPE_CHECKING:
```
is this `if` needed?

enaydanov replied:
Here it's just a habit to divide a stuff used just for type checks only from the runtime stuff. It works together with `from __future__ import annotations`.

But in general, this construction is for avoiding circular imports sometimes required by type hints.

Little bit more here: https://vickiboykis.com/2023/12/11/why-if-type_checking/ (for example)

Also, see https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

kbr-scylla replied:
If conditional is not strictly needed (it will work without it), I'd remove it. And here IIUC there are no circular imports so we can get rid of it

enaydanov replied:
Think of this as a good habit: if some typing stuff is not required for runtime then it should be excluded by this check.

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+LOGGER = logging.getLogger(__name__)
+
+
+if TYPE_CHECKING:
```
is this `if` needed?

enaydanov replied:
see above

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+#   1. Run the prepare part (before the first yield)
+#       >>> await anext(cluster_event_steps)
+#
+#   2. Run the prepare part (before the first yield)
```
Steps 1 and 2 are the same?

kbr-scylla commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,212 @@
+from test.topology_random_failures.cluster_events import CLUSTER_EVENTS, TOPOLOGY_TIMEOUT
+from test.topology_random_failures.error_injections import ERROR_INJECTIONS
+
+if TYPE_CHECKING:
```
is this `if` needed?

enaydanov replied:
see above

kbr-scylla commented on test/topology_random_failures/conftest.py:
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (C) 2023-present ScyllaDB
```
2024

kbr-scylla commented on test/topology_random_failures/error_injections.py:
```diff
@@ -0,0 +1,29 @@
+#
+# Copyright (C) 2023-present ScyllaDB
```
2024

kbr-scylla commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,212 @@
+# Copyright (C) 2023-present ScyllaDB
```
2024

kbr-scylla commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,212 @@
+    servers = [
+        (await manager.server_add(config={
+            "error_injections_at_startup": [
+                {"name": "raft_server_snapshot_reduce_threshold", "one_shot": True}
```
This error injection was removed in 922e03562948c5833a6e37547b2e57b119602ae7 and replaced with more generic one

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+
+@deselect_for(
+    # TODO: remove this skip when #16317 will be resolved.
+    reason="Cannot create CDC log for tables, because keyspace uses tablets. See issue #16317",
```
So the test always runs in tablets mode?

kbr-scylla replied:
We should test CDC in non-tablets mode too. (Especially because CDC in non-tablets mode is already working)

kbr-scylla replied:
Enabling tablets could be a randomly chosen parameter BTW

enaydanov replied:
We plan to extend number of tests. It's a good candidate for another dimension.

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+        "stop_before_streaming",
+        "stop_after_streaming",
+    ],
+    reason="See issue #19151 (decommission process stuck while boostrapping node is paused)",
```
#19151 describes expected behavior. So the event or test should probably get adjusted somehow to account for that.

enaydanov replied:
We decided to leave it outside of this PR.

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+        "stop_before_streaming",
+        "stop_after_streaming",
+    ],
+    reason="See issue #18640 (failed to add a node to a cluster if another bootstrapping node is stuck)",
```
#18640 describes expected behavior. So the event or test should probably get adjusted somehow to account for that.

enaydanov replied:
We decided to leave it outside of this PR.

kbr-scylla commented on test/topology_random_failures/cluster_events.py:
```diff
@@ -0,0 +1,624 @@
+        "stop_before_streaming",
+        "stop_after_streaming",
+    ],
+    reason="See issue #19151 (decommission process stuck while bootstrapping node is paused)",
```
As above

enaydanov replied:
We decided to leave it outside of this PR.

kbr-scylla commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,212 @@
+
+    server_log = await manager.server_open_log(server_id=s_info.server_id)
+
+    if cluster_event_duration + 1 >= WAIT_FOR_IP_TIMEOUT and error_injection in (  # give one more second for a tolerance
```
I smell flakiness

enaydanov replied:
Do you have any ideas how to make it more bulletproof?

kbr-scylla replied:
Can you explain the reasoning behind this logic first? Then I'll try to come up with something.

enaydanov replied:
This code handle a situation when the added node sent a join request and then do not respond for 30+ second (which is wait_for_ip timeout) In this case the new node just hang and we need shut down it manually.

So, I just measure the time spent on a cluster event and if it close to 30 seconds I check coordinator's logs for the message about the node.

kbr-scylla replied:
Hm, ok I think I understand.

If the cluster event took 30+ seconds for whatever reason, then there is a **high chance** (but not a guarantee) that the topology coordinator timed out inside "wait for IP" and this will cause the booting node to hang forever after we unpause it.

But it could be due to process scheduling or whatever, that the topology coordinator didn't yet timeout, but will in about 1 second, and we will miss it the fact. And then we will hang here and timeout and fail the test
```python3
        LOGGER.info("Wait until the new node initialization completes or fails.")
        await server_log.wait_for("init - (Startup failed:|Scylla version .* initialization completed)", timeout=120)
```

So here's alternative proposal that should not depend on timing:
let's **concurrently** wait for one of the following events to happen:
- `server_log` prints `"init - (Startup failed:|Scylla version .* initialization completed)"`
- OR `coordinator_log` prints `"The node may hang ..."`

i.e. you spawn concurrent `wait_for` tasks and then wait until first of them finishes.

And then check if it was the coordinator "node may hang", in this case you can shut down the node.

You can also verify that if this happens, the injections was one of the two listed below (if it's a different injection, then that's unexpected behavior and we should fail the test.)

enaydanov replied:
OK, got your idea. For now test looks pretty stable, but on following up work on it I will try to improve this part too.

kbr-scylla commented on test/topology_random_failures/test_random_failures.py:
```diff
@@ -0,0 +1,212 @@
+    if await manager.running_servers(s_info.server_id):
+        LOGGER.info("Wait until the new node initialization completes or fails.")
+        await server_log.wait_for("init - (Startup failed:|Scylla version .* initialization completed)", timeout=120)
+        time.sleep(1)
```
Why is this sleep needed?

enaydanov replied:
Unfortunately, yes. The scylla process is not shutdown immediately after `Startup failed` message appears in the log. I saw flakiness in my runs because of that.

kbr-scylla replied:
So instead of sleeping for arbitrary time, why not `wait_for` the process to be dead?

enaydanov replied:
In general, it's because `server_log.wait_for()` doesn't return matched line:

https://github.com/scylladb/scylladb/blob/fe0e9618566e02a0128c3247ff3b11c5998b99ea/test/pylib/log_browsing.py#L42-L75

kbr-scylla replied:
so you can either modify `wait_for` to have similar interface to `grep` (which returns matches), OR after `wait_for` returns, call `grep` and it will return the match. Then if it's "failed", `wait_for` process to be dead.

`sleep(1)` is asking for flakiness

enaydanov replied:
OK, replaced it with additional log grep.

enaydanov commented on test/pylib/scylla_cluster.py:
```diff
@@ -518,10 +517,10 @@ async def cql_is_up(self) -> CqlUpState:
         except (NoHostAvailable, InvalidRequest, OperationTimedOut) as exc:
             self.logger.debug("Exception when checking if CQL is up: %s", exc)
-            return CqlUpState.CONNECTED if connected else CqlUpState.NOT_CONNECTED
+            return ServerState.CONNECTED if connected else ServerState.HOST_ID_REQUIRED
```
![Screenshot_20240821_192502](https://github.com/user-attachments/assets/d126bb6b-771b-4994-ac86-04172fbb151c)

kbr-scylla replied:
Ok, but let's make the naming consistent, accurate, and remove abstraction leak:
- `HOST_ID_REQUIRED` should be `HOST_ID_QUERIED`. This is the state we enter **after** querying host ID. (see also my other comments below where the states are assigned) "host ID required" is really the same as "process started" so we don't need it
- I'd rename `get_server_state` back to `cql_is_up` because this function really is only dealing with CQL connectivity.
- rename `ServerState` -> `ServerUpState` to be more precise
- rename `CONNECTED` -> `CQL_CONNECTED` to be more precise
- now `cql_is_up` should not be aware of "host ID", the fact that it returns `HOST_ID_<something>` in some cases is an abstraction leak. To get rid of it, I'd introduce a parameter to this function, `start_state`, and if we fail to connect, return `start_state`. At the callsite we would pass `start_state` to be the state we're currently in (it will be `HOST_ID_REQUIRED`) (see comment below)
- alternatively if you don't like passing `start_state` to `cql_is_up`, we could make `cql_is_up` return `ServerUpState | None` -- if it fails to connect, return `None`. Then at call site we would do `server_up_state = await self.cql_is_up() or server_up_state` (which means: if we successfully connect with CQL, assign the new state after connecting; if we fail, we remain in whatever state we had (which would be `HOST_ID_REQUIRED`)). I think I actually like this idea more than passing `start_state`, but choose your favorite

enaydanov replied:
OK, will go with your latest idea. One thing to improve naming here is to use name `get_cql_up_state` instead of `cql_is_up` which sounds like it should return boolean value.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11237/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11237/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
#### Build Details:
- Duration: 51 min
- Builder: spider2.cloudius-systems.com

enaydanov commented on test/pylib/scylla_cluster.py:
```diff
@@ -582,20 +586,24 @@ def report_error(message: str):
         while time.time() < self.start_time + self.TOPOLOGY_TIMEOUT and not self.stop_event.is_set():
             assert self.cmd is not None
             if self.cmd.returncode:
+                server_state = ServerState.PROCESS_EXITED
```
Actually, this line is redundant. We use `expected_error` argument to handle expected scylla process exits.

enaydanov commented:
> > Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2442/testReport/(root)/test_random_failures/
> 
> Does the PR description and linked sheet have to be updated?
> 
> In fact the sheet should probably not be included in the PR description, or the known issues -- all of that is going to be made permanent part of Scylla's `git log`.

Everything is updated and this part of PR description is not a part of the commit messages.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11241/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11241/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11241/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11354/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/284/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3331/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11241/)
#### Build Details:
- Duration: 4 hr 14 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11267/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11267/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11267/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/305/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3352/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11376/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11267/)
#### Build Details:
- Duration: 4 hr 2 min
- Builder: spider7.cloudius-systems.com

enaydanov commented:
Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2465/testReport/(root)/test_random_failures/

enaydanov commented:
@kbr-scylla can you look again?

kbr-scylla commented on test/pylib/scylla_cluster.py:
```diff
@@ -575,14 +577,16 @@ async def start(self, api: ScyllaRESTAPIClient, expected_error: Optional[str] =
         self.start_time = time.time()
         sleep_interval = 0.1
-        cql_up_state = CqlUpState.NOT_CONNECTED
+        server_state = ServerState.HOST_ID_REQUIRED
```
We should start with `PROCESS_STARTED` here. See also below

kbr-scylla replied:
and rename `server_state` -> `server_up_state` (more accurate)

kbr-scylla commented on test/pylib/scylla_cluster.py:
```diff
@@ -601,18 +605,21 @@ def report_error(message: str):
                         report_error("the node startup failed, but the log file doesn't contain the expected error")
                 report_error("failed to start the node")
 
             if hasattr(self, "host_id") or await self.get_host_id(api):
```
After this line we should assign `server_up_state = HOST_ID_QUERIED`
and then (see comments above)  `server_up_state = await self.cql_is_up(..., start_state=server_up_state)`

kbr-scylla commented on test/topology_random_failures/suite.yaml:
```diff
@@ -0,0 +1,11 @@
+type: Topology
```
Do we need another suite? Can't we reuse `topology_custom`? It also gives a 0 size cluster.

enaydanov replied:
Yes, it's better to use a separate suite here because of own conftest and some python modules which are for this test only.

kbr-scylla commented:
> Everything is updated and this part of PR description is not a part of the commit messages.

Yes but we create merge commit when merging, using PR description. So it will end up in `git log`

enaydanov commented:
Known issues:

- https://github.com/scylladb/scylladb/issues/16317
- https://github.com/scylladb/scylladb/issues/18068
- https://github.com/scylladb/scylladb/issues/18640
- https://github.com/scylladb/scylladb/issues/19151

Test matrix:

[![Test matrix](https://github.com/user-attachments/assets/12ed8afe-c7b5-47e4-82bc-ecca28b385eb)](https://docs.google.com/spreadsheets/d/1GcZuEAUlvP4KtgHNc_geBiirqE46yZtrb9lOev5--4E/edit#gid=0)

enaydanov commented:
Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2466/testReport/(root)/test_random_failures/

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11309/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11309/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11309/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3382/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/333/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11404/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11309/)
#### Build Details:
- Duration: 8 hr 24 min
- Builder: i-0b072d86cfa8910b2 (m5ad.8xlarge)

enaydanov commented:
@kbr-scylla it is ready for next round. Please, review.

kostja reviewed (approved):
Aside from agreeing with @kbr-scylla and a bunch of items that are already left for a follow up I have nothing more to add.

enaydanov commented:
@kostja @kbr-scylla 

I've removed `--raft-*` CLI arguments and added constants to `test_random_failures.py` instead, also I've renamed Raft topology version of `stop_after_setting_mode_to_normal` to `stop_after_setting_mode_to_normal_raft_topology` and removed the gossip version of it because I don't use it in this test.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11349/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11349/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11349/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3404/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/355/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11427/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11349/)
#### Build Details:
- Duration: 3 hr 55 min
- Builder: spider1.cloudius-systems.com

enaydanov commented:
Full matrix passed: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/2467/testReport/(root)/test_random_failures/

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11383/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11383/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_random_failures/test_random_failures
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11383/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/382/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3432/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11455/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11383/)
#### Build Details:
- Duration: 4 hr 28 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
@enaydanov is it ready for merging?

kbr-scylla commented:
Qd
